### PR TITLE
feat: remove the need to call AppError when creating a Report Error

### PR DESCRIPTION
**Pull request by VitorCaminha**
remove the need to call a instance of AppError in the arguments when creating a new Report Error, replacing it with a single object argument, receiving the status code, message and optional service name, making the user code more clean and easy to read.

In future, we could create new methods in Report class to register debug or info logs aswell.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -91,14 +91,14 @@ export const UserContainerModule = CreateModule([
...
--   **Error handling**: Using `Report.Error()` we can report errors using predefined error codes. This is useful when we want to report known errors to the client. Inside of `Report.Error()` there is a try catch block encapsulating the error.
+-   **Error handling**: Using `new Report().Error()` we can report errors using predefined error codes. This is useful when we want to report known errors to the client. Inside of `new Report().Error()` there is a try catch block encapsulating the error.
 
 ```typescript
 type ErrorType = GeneralErrorCode | ApplicationErrorCode | HttpStatusErrorCode;
 
 // Reporting errors
 if (userExist) {
-    const error = Report.Error(new ApplicationError(StatusCode.BadRequest, "User already exist!"), "user-create");
+    const error = new Report().Error({ statusCode: StatusCode.BadRequest, message: "User already exist!", service: "user-create"});
```
Comment (rsaz) [inline README.md:101]: The initial is great, however the user experience using new Report().Error is still a bit odd. The only problem we actually need to solve is that how we can capture the stack trace from where the error happen without constructing an instance of a type?
If we achieve this we can supply to the user a smooth experience like Report.Error(statusCode.BadRequest, "User already created", "user-create-usecase")

File: packages/core/src/error/error-handler-middleware.ts [hunk 2]
```diff
@@ -1,9 +1,9 @@
 import { NextFunction, Request, Response } from "express";
-import { AppError } from "./application-error";
 import { StatusCode } from "./status-code";
 import { LogLevel, log } from "../logger";
+import { IAppError } from "./report";
 
-function errorHandler(error: AppError, req: Request, res: Response, next: NextFunction): void {
+function errorHandler(error: IAppError, req: Request, res: Response, next: NextFunction): void {
     
     log(LogLevel.Error, error, error.service || "service-undefined");
```
Comment (rsaz) [inline packages/core/src/error/error-handler-middleware.ts:8]: We also need to change the service-undefined. How can we find the service that generate the error?

File: packages/core/src/error/report.ts [hunk 3]
```diff
@@ -1,14 +1,22 @@
...
+    constructor() {
+        super()
+    }
 
-    public static Error(error: AppError) {
-    
-        throw error;
+    public Error(appError: IAppError) {
+        
+        throw { ...appError, name: this.name, stack: this.stack };
```
Comment (rsaz) [inline packages/core/src/error/report.ts:17]: Preferable having a type rather than ducktype {}

**Review by rsaz — CHANGES_REQUESTED** [review]
You are in the right track. Try to evaluate how other languages resolve this problem as comparison

File: README.md (hunk 1 above)
Comment (VitorCaminha) [reply README.md]: Done!

**Comment by rsaz** [comment]
This item will be merged in the next release. @VitorCaminha

**Comment by rsaz** [comment]
Solution merged in the `feature/error-handling` branch, commit: `. Thanks for the submission.

**Pull request closed without merging**